File: frontend/src/domain/temporal.ts
/**
 * Temporal data-state model - the frontend mirror of the backend
 * `app.market_data.temporal` vocabulary (Step 13C). The dashboard fallback resolver stamps
 * each value's origin so a UI cell can answer: what is the value, which session is it from,
 * when was it observed, is it live vs last-session vs unavailable.
 */

export type DataTemporalState =
  "LIVE" | "LAST_SESSION" | "HISTORICAL" | "DERIVED" | "UNAVAILABLE";

export type DisplayState = "LIVE" | "LAST_SESSION" | "MIXED" | "UNAVAILABLE";

export interface FieldProvenance {
  state: DataTemporalState;
  source: string;
  asOf?: string | null; // ISO8601 (VN)
  sessionDate?: string | null; // VN trading-session date
  stale?: boolean;
  note?: string | null;
}

export interface RowProvenance {
  quote: FieldProvenance;
  book: FieldProvenance;
  analytics?: FieldProvenance;
}

export function quoteTimestamp(
  q?: {
    exchangeTimestamp?: number | null;
    sourceTimestamp?: number | null;
  } | null,
): string | null {
  const ts = q?.exchangeTimestamp ?? q?.sourceTimestamp;
  if (typeof ts !== "number" || !Number.isFinite(ts) || ts <= 0) return null;
  const value = new Date(ts);
  return Number.isFinite(value.getTime()) ? value.toISOString() : null;
}

export function formatAsOf(iso?: string | null): string {
  if (!iso) return "Time unavailable";
  if (iso.length === 10) {
    const day = new Date(`${iso}T00:00:00+07:00`);
    return Number.isFinite(day.getTime())
      ? new Intl.DateTimeFormat("en-GB", {
          timeZone: "Asia/Ho_Chi_Minh",
          day: "numeric",
          month: "short",
          year: "numeric",
        }).format(day) + " · session date (ICT)"
      : "Time unavailable";
  }
  const d = new Date(iso);
  if (!Number.isFinite(d.getTime())) return "Time unavailable";
  return (
    new Intl.DateTimeFormat("en-GB", {
      timeZone: "Asia/Ho_Chi_Minh",
      day: "numeric",
      month: "short",
      hour: "2-digit",
      minute: "2-digit",
      second: "2-digit",
      hour12: false,
    }).format(d) + " ICT"
  );
}

/** Short label for a display state (no styling opinion - the visual phase owns that). */
export function temporalLabel(state: DisplayState): string {
  switch (state) {
    case "LIVE":
      return "Live";
    case "LAST_SESSION":
      return "Last session";
    case "MIXED":
      return "Live / last session";
    case "UNAVAILABLE":
    default:
      return "Unavailable";
  }
}

/** "as of Fri 28 Aug" style helper from an ISO instant or a session date. */
export function asOfLabel(p?: FieldProvenance): string | null {
  const iso = p?.asOf || p?.sessionDate;
  if (!iso) return null;
  const d = new Date(iso.length <= 10 ? `${iso}T15:00:00+07:00` : iso);
  if (isNaN(d.getTime())) return null;
  return d.toLocaleDateString("en-GB", {
    weekday: "short",
    day: "numeric",
    month: "short",
    timeZone: "Asia/Ho_Chi_Minh",
  });
}

File: backend/app/market_data/temporal.py
"""Canonical temporal data-state model (Step 13C).

Every value the terminal displays carries exactly one temporal origin. The UI must be able
to answer, for any cell: what is the value, which trading session is it from, when was it
observed, is it live, is it last-session, is it unavailable and why. This module is the
vocabulary; the resolver (`market_snapshot_resolver`) assigns it.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any


class DataTemporalState(str, Enum):
    LIVE = "LIVE"                  # fresh tick during an active session
    SESSION_SNAPSHOT = "SESSION_SNAPSHOT"  # current-session value while matching is paused
    LAST_SESSION = "LAST_SESSION"  # final / most-recent value from the last completed session
    HISTORICAL = "HISTORICAL"      # persisted daily bar older than the last completed session
    DERIVED = "DERIVED"            # computed from other stated values (e.g. change vs prior close)
    UNAVAILABLE = "UNAVAILABLE"    # realtime-only field with no legitimate fallback


class DataSource(str, Enum):
    LIVE_FEED = "LIVE_FEED"                  # FiinQuant SignalR -> MarketState
    REDIS_WARM = "REDIS_WARM"                # warm-cache restore (still same-session)
    SNAPSHOT_FINAL = "SNAPSHOT_FINAL"        # instrument_snapshots, quality=FINAL (session close)
    SNAPSHOT_CHECKPOINT = "SNAPSHOT_CHECKPOINT"  # instrument_snapshots, mid-session checkpoint
    SNAPSHOT_SEED = "SNAPSHOT_SEED"          # instrument_snapshots seeded from EOD bars (no book)
    EOD_BARS = "EOD_BARS"                    # market_bars daily close/OHLC/volume
    PRIOR_CLOSE = "PRIOR_CLOSE"              # previous trading session's close (used as reference)
    SESSION_REFERENCE = "SESSION_REFERENCE"  # provider snapshot: prior close + session bands
    QUANT_LIVE = "QUANT_LIVE"               # LiveQuantEngine, current session
    QUANT_EOD = "QUANT_EOD"                 # LiveQuantEngine recomputed for a completed session
    NONE = "NONE"                            # no source


# Row-level badge the UI can render without inspecting every field group.
class DisplayState(str, Enum):
    LIVE = "LIVE"
    LAST_SESSION = "LAST_SESSION"
    MIXED = "MIXED"              # some groups live, some last-session (e.g. transition minute)
    UNAVAILABLE = "UNAVAILABLE"  # nothing legitimate to show
    SESSION_SNAPSHOT = "SESSION_SNAPSHOT"


@dataclass(frozen=True)
class FieldProvenance:
    """Temporal origin of one *group* of related fields (quote / book / analytics)."""

    state: DataTemporalState
    source: DataSource
    as_of: str | None = None        # ISO8601 (VN) instant the value was observed/computed
    session_date: str | None = None  # VN trading-session date the value belongs to
    stale: bool = False              # true when the value is older than the last completed session
    note: str | None = None          # short human reason, esp. for UNAVAILABLE

    def to_wire(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "state": self.state.value,
            "source": self.source.value,
        }
        if self.as_of is not None:
            d["asOf"] = self.as_of
        if self.session_date is not None:
            d["sessionDate"] = self.session_date
        if self.stale:
            d["stale"] = True
        if self.note:
            d["note"] = self.note
        return d


UNAVAILABLE_QUOTE = FieldProvenance(DataTemporalState.UNAVAILABLE, DataSource.NONE)


def derive_display_state(groups: list[FieldProvenance]) -> DisplayState:
    states = {g.state for g in groups if g.state != DataTemporalState.UNAVAILABLE}
    if not states:
        return DisplayState.UNAVAILABLE
    if states == {DataTemporalState.LIVE}:
        return DisplayState.LIVE
    if states == {DataTemporalState.SESSION_SNAPSHOT}:
        return DisplayState.SESSION_SNAPSHOT
    if DataTemporalState.LIVE in states:
        return DisplayState.MIXED
    return DisplayState.LAST_SESSION
